extensions/PhabBugz/lib/Repository.pm

```diff
@@ -0,0 +1,121 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+# This Source Code Form is "Incompatible With Secondary Licenses", as
+# defined by the Mozilla Public License, v. 2.0.
+
+package Bugzilla::Extension::PhabBugz::Repository;
+
+use 5.10.1;
+use Moo;
+use Types::Standard -all;
+
+use Bugzilla::Util qw(trim);
+use Bugzilla::Extension::PhabBugz::Util qw(request);
+
+#########################
+#    Initialization     #
+#########################
+
+has id              => (is => 'ro',   isa => Int);
+has phid            => (is => 'ro',   isa => Str);
+has type            => (is => 'ro',   isa => Str);
+has name            => (is => 'ro',   isa => Str);
+has description     => (is => 'ro',   isa => Maybe [Str]);
+has vcs             => (is => 'ro',   isa => Str);
+has status          => (is => 'ro',   isa => Str);
+has call_sign       => (is => 'ro',   isa => Str);
+has short_name      => (is => 'ro',   isa => Str);
+has default_branch  => (is => 'ro',   isa => Str);
+has creation_ts     => (is => 'ro',   isa => Str);
+has modification_ts => (is => 'ro',   isa => Str);
+has view_policy     => (is => 'ro',   isa => Str);
+has edit_policy     => (is => 'ro',   isa => Str);
+has push_policy     => (is => 'ro',   isa => Str);
+
+sub new_from_query {
+  my ($class, $params) = @_;
+
+  my $data
+    = {queryKey => 'all', constraints => $params};
+
+  my $result = request('diffusion.repository.search', $data);
```

Comment:
Good point. If we have a stack of different revisions bu they all share the same repository, then we would do a new lookup for each revision. I will update this to use request_cache so we would only lookup a repo once for a page request.

Reply:
> Good point. If we have a stack of different revisions bu they all share the same repository, then we would do a new lookup for each revision. I will update this to use request_cache so we would only lookup a repo once for a page request.

Sounds good. Looks like `_build_repository` will cache known repositories between requests too, right? That would be nice to have so we can check the cache for know repos and just return the result from there instead of hitting Phab.

Reply:
No. request_cache is only for a single page request so the cached objects would be removed for another page. But we really don't want them to stick around on case the repo information changes on Phabricator and we have no way to invalidate the cache on BMO if that happens. I feel it is acceptable to load the repo once for a page as it is happening in the background and is not noticeable really for single call.